ocfweb/account/vhost_mail.py

```diff
@@ -117,6 +135,137 @@ def vhost_mail_update(request):
+def vhost_mail_csv_export(request, domain):
+    user = logged_in_user(request)
+    vhost = _get_vhost(user, domain)
+    if not vhost:
+        _error(request, 'You cannot use the domain: "{}"'.format(domain))
+
+    with _txn() as c:
+        addresses = (
+            addr for addr in sorted(vhost.get_forwarding_addresses(c))
+            if not addr.is_wildcard
+        )
+
+    response = HttpResponse(
+        _write_csv(addresses),
+        content_type='text/csv',
+    )
+    # See https://docs.djangoproject.com/en/1.11/ref/request-response/
+    response['Content-disposition'] = 'attachment; filename="{}.csv"'.format(domain)
+    return response
+
+
+@login_required
+@group_account_required
+@require_POST
+def vhost_mail_csv_import(request, domain):
+    user = logged_in_user(request)
+    vhost = _get_vhost(user, domain)
+    if not vhost:
+        _error(request, 'You cannot use the domain: "{}"'.format(domain))
+
+    addresses = _parse_csv(request, domain)
+
+    # Add new addresses and update existing if the recipients changed
+    with _txn() as c:
+        existing_addrs = {
+            addr.address: addr
+            for addr in vhost.get_forwarding_addresses(c)
+        }
+
+        for from_addr, to_addrs in addresses.items():
+            if from_addr in existing_addrs:
+                existing_addr = existing_addrs[from_addr]
+                if to_addrs != existing_addr.forward_to:
+                    new = MailForwardingAddress(
+                        address=from_addr,
+                        crypt_password=existing_addr.crypt_password,
+                        forward_to=to_addrs,
+                        last_updated=None,
+                    )
+                    vhost.remove_forwarding_address(c, from_addr)
+                else:
+                    new = None
+            else:
+                new = MailForwardingAddress(
+                    address=from_addr,
+                    crypt_password=None,
+                    forward_to=to_addrs,
+                    last_updated=None,
+                )
+
+            if new is not None:
+                vhost.add_forwarding_address(c, new)
+
+    messages.add_message(request, messages.SUCCESS, 'Import successful!')
+    return _redirect_back()
+
+
+def _write_csv(addresses):
+    """Turn a collection of vhost forwarding addresses into a CSV
+    string for user download."""
+    buf = io.StringIO()
+    writer = csv.writer(buf)
+    for addr in addresses:
+        writer.writerow((
+            addr.address.split('@')[0],
+            ' '.join(sorted(addr.forward_to)),
+        ))
+
+    return buf.getvalue()
+
+
+def _parse_csv(request, domain):
+    """Parse, validate, and return addresses from the file uploaded
+    with the CSV upload button/form."""
+    csv_file = request.FILES.get('csv_file')
+    if not csv_file:
+        _error(request, 'Missing CSV file!')
+
+    addresses = {}
+    with io.TextIOWrapper(csv_file, encoding='utf-8') as f:
```

Comment:
Let's do something more graceful when someone gives us non-UTF-8 text than crash:
![2017-09-30 02-57-39](https://user-images.githubusercontent.com/4122890/31044808-7f382bee-a58b-11e7-9946-1f3f36a6484e.png)

For that matter, since we don't accept special characters anyway, we could as well use ascii for our codec.

Reply:
That's a good idea, it should display an error message. That's probably more helpful than simply stating "UTF-8 encoded" on the page anyways.

I don't think there's any need to restrict it to ASCII, though. We don't support unicode now, but I don't want to add barriers for anyone who wants to add support later.

For the record, I believe postfix supports UTF-8, so the validation regex and perhaps the database might be the only hindrances to someone wanting to test that out.